File: max/engine/blend.js
"use strict";

/**
 * blend.js — cross-corpus blending: Color (corpus morph) + Adventure
 * (temperature) + Gravity (cadence bias).
 *
 * Direct port of python/src/blend.py. Two performable dials act on the per-corpus
 * transition distributions:
 *   - Color c in [0,1] crossfades along an ordered path of single-corpus anchors
 *     (COLOR_PATH). Piecewise-linear, so anchors are pure single corpora and
 *     positions between them are genuine two-corpus blends.
 *   - Adventure a in [0,1] sets a sampling temperature tau: low sharpens toward
 *     the likeliest chord, high flattens the tail so rarer chords surface.
 * A third dial, Gravity, biases the result toward the tonic/dominant of the mode.
 *
 * blendedChoices returns Array<[target, prob]> ready for weighted sampling, or []
 * when the source chord is absent from every weighted corpus (the caller then
 * falls back to the pooled "all" chain / fallback policy).
 *
 * entries throughout are Array<[target, prob]> (the JS shape corpus.js produces).
 * Ordering is significant: sorts are DESCENDING by prob and STABLE, so ties keep
 * the incoming order — this reproduces Python's `sorted(key=lambda kv: -kv[1])`
 * over a dict built in a fixed insertion order. The parity fixture pins this.
 */

const { PITCH_CLASSES, keyOffset, transposeChord } = require("./chord_vocab.js");

// Ordered single-corpus anchors for the Color dial (config.COLOR_PATH):
// 0.0 = plainest (folk) ... 1.0 = spiciest (jazz).
const COLOR_PATH = ["nottingham", "pop909", "bach", "openbook"];

// Adventure -> tau linear interpolation bounds (config.ADVENTURE_TAU_*).
const ADVENTURE_TAU_MIN = 0.6;
const ADVENTURE_TAU_MAX = 1.8;

// Cadence / harmonic-gravity constants (config.*). Roots are fixed in the
// normalized C/Am space: C/G for major, A/E for minor.
const TONIC_PC = { maj: 0, min: 9 };
const DOMINANT_PC = { maj: 7, min: 4 };
const CADENCE_TONIC_BOOST = 3.0;
const CADENCE_DOMINANT_BOOST = 1.2;

function clamp01(x) {
  return x < 0 ? 0 : x > 1 ? 1 : x;
}

/**
 * colorWeights(c, available) -> Map<name, weight> summing to 1.
 * Anchors absent from `available` are dropped from the path first, so the dial
 * still spans whatever corpora actually loaded. `available` null/undefined keeps
 * the whole path.
 */
function colorWeights(c, available) {
  const path = COLOR_PATH.filter(
    (name) => available == null || available.indexOf(name) !== -1
  );
  if (path.length === 0) return new Map();
  if (path.length === 1) return new Map([[path[0], 1.0]]);

  c = clamp01(c);
  const span = path.length - 1;
  const pos = c * span;
  // Segment index; clamp so c==1 stays inside the last segment (frac -> 1).
  const i = Math.min(Math.floor(pos), span - 1);
  const frac = pos - i;
  const weights = new Map();
  for (const name of path) weights.set(name, 0.0);
  weights.set(path[i], weights.get(path[i]) + (1.0 - frac));
  weights.set(path[i + 1], weights.get(path[i + 1]) + frac);
  // Drop zero-weight anchors, preserving path order.
  const out = new Map();
  for (const [k, v] of weights) if (v > 0.0) out.set(k, v);
  return out;
}

/** temperature(a) -> tau. Linear: 0.6 + a*(1.8-0.6). */
function temperature(a) {
  a = clamp01(a);
  return ADVENTURE_TAU_MIN + a * (ADVENTURE_TAU_MAX - ADVENTURE_TAU_MIN);
}

/**
 * applyTemperature(entries, tau) -> reshaped entries, renormalized, SORTED
 * DESCENDING by prob. Reshape is p**(1/tau) over strictly-positive probs; a
 * non-positive tau is floored to 1e-3 (mirrors Python). Returns [] if the
 * reshaped mass is non-positive.
 */
function applyTemperature(entries, tau) {
  if (tau <= 0) tau = 1e-3;
  const inv = 1.0 / tau;
  const reshaped = []; // [target, p**inv], insertion order preserved
  let total = 0;
  for (const [t, p] of entries) {
    if (p > 0.0) {
      const v = Math.pow(p, inv);
      reshaped.push([t, v]);
      total += v;
    }
  }
  if (total <= 0) return [];
  const out = reshaped.map(([t, v]) => [t, v / total]);
  // Descending, stable: ties keep the reshaped (insertion) order, matching
  // Python's `sorted(key=lambda kv: -kv[1])`.
  out.sort((a, b) => b[1] - a[1]);
  return out;
}

/** Pitch class of a 'Root:quality' symbol's root, or null for non-chords. */
function rootPc(chord) {
  const i = chord.indexOf(":");
  const root = i === -1 ? chord : chord.slice(0, i); // partition(":")[0]
  const pc = PITCH_CLASSES[root];
  return pc === undefined ? null : pc;
}

/**
 * applyCadence(entries, mode, gravity) -> entries biased toward the tonic (and
 * secondarily the dominant) of `mode`, renormalized and re-sorted descending.
 * gravity <= 0 (or empty entries) is identity. Assumes entries are in normalized
 * C/Am space so the tonic/dominant roots are fixed per mode.
 */
function applyCadence(entries, mode, gravity) {
  if (gravity <= 0 || entries.length === 0) return entries;
  const tonic = mode in TONIC_PC ? TONIC_PC[mode] : TONIC_PC.maj;
  const dom = mode in DOMINANT_PC ? DOMINANT_PC[mode] : DOMINANT_PC.maj;
  const boosted = [];
  let total = 0;
  for (const [target, prob0] of entries) {
    let prob = prob0;
    const pc = rootPc(target);
    if (pc === tonic) prob *= 1.0 + gravity * CADENCE_TONIC_BOOST;
    else if (pc === dom) prob *= 1.0 + gravity * CADENCE_DOMINANT_BOOST;
    boosted.push([target, prob]);
    total += prob;
  }
  if (total <= 0) return entries;
  const out = boosted.map(([t, p]) => [t, p / total]);
  out.sort((a, b) => b[1] - a[1]);
  return out;
}

/**
 * blendedChoices(corpora, weights, tau, sourceChord, mode, gravity)
 *   -> Array<[target, prob]> (empty if sourceChord unknown in all weighted corpora).
 *
 * For each weighted corpus that CONTAINS sourceChord: mixed[target] += w * prob.
 * Then divide every value by total_weight (summed weight of ONLY the corpora that
 * had it), apply temperature, then cadence. sourceChord must already be in
 * normalized (C/Am) key space.
 */
function blendedChoices(corpora, weights, tau, sourceChord, mode, gravity) {
  mode = mode === undefined ? "maj" : mode;
  gravity = gravity === undefined ? 0.0 : gravity;

  const mixed = new Map(); // target -> accumulated w*prob, first-seen order
  let totalWeight = 0.0;
  // Iterate weights in insertion order (color_weights path order) so the mixed
  // insertion order — and thus post-sort tie-breaking — matches Python.
  for (const [name, w] of weights) {
    if (w <= 0) continue;
    if (!corpora.has(name, sourceChord)) continue; // present & non-empty
    const dist = corpora.distBySource(name).get(sourceChord);
    totalWeight += w;
    for (const [target, prob] of dist) {
      mixed.set(target, (mixed.get(target) || 0.0) + w * prob);
    }
  }

  if (mixed.size === 0 || totalWeight <= 0) return [];

  // Renormalize by the summed weight of only the corpora that contributed.
  const norm = [];
  for (const [t, p] of mixed) norm.push([t, p / totalWeight]);
  return applyCadence(applyTemperature(norm, tau), mode, gravity);
}

/**
 * normalizeToKey(chord, key) -> [normalizedChord, offset].
 * Transpose an in-key chord into normalized (C/Am) space; apply -offset to get
 * back to the key.
 */
function normalizeToKey(chord, key) {
  const offset = keyOffset(key);
  return [transposeChord(chord, offset), offset];
}

module.exports = {
  COLOR_PATH,
  colorWeights,
  temperature,
  applyTemperature,
  applyCadence,
  blendedChoices,
  normalizeToKey,
};

File: versions/v4-theory/python/src/blend.py
"""Cross-corpus blending: Color (corpus morph) + Adventure (temperature).

The two performable dials both act on the per-corpus transition distributions:

* **Color** ``c in [0,1]`` morphs a 1-D crossfade along an ordered path of
  single-corpus anchors (``COLOR_PATH``, e.g. folk -> pop -> classical -> jazz).
  Piecewise-linear, so anchor positions are pure single corpora and in-between
  positions are genuine two-corpus blends; every corpus is reachable.
* **Adventure** ``a in [0,1]`` sets a sampling temperature ``tau``: low sharpens
  toward the likeliest chord (safe), high flattens the tail so rarer, more
  colorful chords surface.

``blended_choices`` returns ``[(target, prob)]`` ready for weighted sampling, or
an empty list when the source chord is absent from every weighted corpus (the
caller then falls back to the pooled ``all`` chain / fallback policy).
"""

from __future__ import annotations

from .config import (
    ADVENTURE_TAU_MAX,
    ADVENTURE_TAU_MIN,
    CADENCE_DOMINANT_BOOST,
    CADENCE_TONIC_BOOST,
    COLOR_PATH,
    DOMINANT_PC,
    TONIC_PC,
)
from .corpus_loader import CorporaSet
from .chord_vocab import PITCH_CLASSES, key_offset, transpose_chord


def _clamp01(x: float) -> float:
    return 0.0 if x < 0 else 1.0 if x > 1 else x


def color_weights(c: float, available: list[str] | None = None) -> dict[str, float]:
    """Corpus mix weights for Color position ``c`` along ``COLOR_PATH``.

    Anchors not present in ``available`` are dropped from the path first, so the
    dial still spans the corpora that actually loaded (e.g. before Bach is
    built). Returns weights summing to 1.
    """
    path = [name for name in COLOR_PATH if available is None or name in available]
    if not path:
        return {}
    if len(path) == 1:
        return {path[0]: 1.0}

    c = _clamp01(c)
    span = len(path) - 1
    pos = c * span
    i = min(int(pos), span - 1)  # segment index; clamp so c==1 stays in last seg
    frac = pos - i
    weights = {name: 0.0 for name in path}
    weights[path[i]] += 1.0 - frac
    weights[path[i + 1]] += frac
    return {k: v for k, v in weights.items() if v > 0.0}


def temperature(a: float) -> float:
    """Adventure position -> sampling temperature tau (linear interpolation)."""
    a = _clamp01(a)
    return ADVENTURE_TAU_MIN + a * (ADVENTURE_TAU_MAX - ADVENTURE_TAU_MIN)


def _apply_temperature(dist: dict[str, float], tau: float) -> list[tuple[str, float]]:
    if tau <= 0:
        tau = 1e-3
    inv = 1.0 / tau
    reshaped = {t: (p ** inv) for t, p in dist.items() if p > 0.0}
    total = sum(reshaped.values())
    if total <= 0:
        return []
    return sorted(((t, p / total) for t, p in reshaped.items()), key=lambda kv: -kv[1])


def _root_pc(chord: str) -> int | None:
    """Pitch class of a ``Root:quality`` symbol's root, or None for non-chords."""
    return PITCH_CLASSES.get(chord.partition(":")[0])


def _apply_cadence(
    choices: list[tuple[str, float]], mode: str, gravity: float
) -> list[tuple[str, float]]:
    """Bias a ``[(target, prob)]`` list toward the tonic/dominant of ``mode``.

    ``choices`` is assumed to be in normalized (C/Am) key space (as produced by
    ``blended_choices``), so the tonic and dominant roots are fixed per mode.
    ``gravity`` 0 -> identity; 1 -> full boost. Renormalizes and keeps the
    descending-probability order the caller expects.
    """
    if gravity <= 0 or not choices:
        return choices
    tonic = TONIC_PC.get(mode, TONIC_PC["maj"])
    dom = DOMINANT_PC.get(mode, DOMINANT_PC["maj"])
    boosted: list[tuple[str, float]] = []
    for target, prob in choices:
        pc = _root_pc(target)
        if pc == tonic:
            prob *= 1.0 + gravity * CADENCE_TONIC_BOOST
        elif pc == dom:
            prob *= 1.0 + gravity * CADENCE_DOMINANT_BOOST
        boosted.append((target, prob))
    total = sum(p for _, p in boosted)
    if total <= 0:
        return choices
    return sorted(((t, p / total) for t, p in boosted), key=lambda kv: -kv[1])


def blended_choices(
    corpora: CorporaSet,
    weights: dict[str, float],
    tau: float,
    source_chord: str,
    mode: str = "maj",
    gravity: float = 0.0,
) -> list[tuple[str, float]]:
    """Mix weighted corpora for ``source_chord``, temper, return [(target, prob)].

    Empty list -> source chord unknown in all weighted corpora (caller falls
    back). ``source_chord`` must already be in normalized (C/Am) key space.
    ``mode``/``gravity`` apply the cadence bias (``gravity=0`` -> no change).
    """
    mixed: dict[str, float] = {}
    total_weight = 0.0
    for name, w in weights.items():
        if w <= 0:
            continue
        table = corpora.corpora.get(name)
        if table is None:
            continue
        dist = table.dist_by_source.get(source_chord)
        if not dist:
            continue
        total_weight += w
        for target, prob in dist.items():
            mixed[target] = mixed.get(target, 0.0) + w * prob

    if not mixed or total_weight <= 0:
        return []

    # Renormalize the mixture (weights of absent corpora were dropped above).
    norm = {t: p / total_weight for t, p in mixed.items()}
    return _apply_cadence(_apply_temperature(norm, tau), mode, gravity)


def normalize_to_key(chord: str, key: str) -> tuple[str, int]:
    """Transpose an in-key chord into normalized (C/Am) space.

    Returns (normalized_chord, offset); apply ``-offset`` to get back to key.
    """
    offset = key_offset(key)
    return transpose_chord(chord, offset), offset
